text/0000-naked-fns.md

```diff
@@ -0,0 +1,186 @@
+Rust, it is advantageous to allow the Rust compiler to drive the entire process,
+particularly in that code may be generated via monomorphization or macro
+expansion.
+
+When writing interrupt handlers for example, most systems require additional
+state be saved beyond the usual ABI requirements.  To avoid corrupting program
+state, the interrupt handler must save the registers which might be modified
+before handing control to compiler-generated code. Consider a contrived
+interrupt handler for x86\_64:
+
+```rust
+unsafe fn isr_nop() {
+    asm!("push %rax"
+         /* Additional pushes elided */ :::: "volatile");
+    let n = 0u64;
+    asm!("pop %rax"
+         /* Additional pops elided */ :::: "volatile");
+}
+```
+
+The generated assembly for this function might resemble the following
+(simplified for readability):
+
+```x86
+isr_nop:
+    sub $8, %rsp
+    push %rax
+    movq $0, 0(%rsp)
+    pop %rax
+    add $8, %rsp
+    retq
+```
+
+Here the programmer's need to save machine state conflicts with the compiler's
+assumption that it has complete control over stack layout, with the result that
+the saved value of `rax` is clobbered by the compiler. Given that details of
+stack layout for any given function are not predictable (and may change with
+compiler version or optimization settings), attempting to predict the stack
+layout to sidestep this issue is infeasible.
+
+When interacting with FFIs that are not natively supported by the compiler,
+a similar situation arises where the programmer knows the expected calling
+convention and can implement a translation between the foreign ABI and one
+supported by the compiler.
+
+Support for naked functions also allows programmers to write functions that
+would otherwise be unsafe, such as the following snippet which returns the
+address of its caller when called with the C ABI on x86.
+
+```
+    mov 4(%ebp), %eax
+    ret
+```
+
+---
+
+Because the compiler depends on a function prologue and epilogue to maintain
+storage for local variable bindings, it is generally unsafe to write anything
+but inline assembly inside a naked function.  The [LLVM language
+reference](http://llvm.org/docs/LangRef.html#function-attributes) describes this
+feature as having "very system-specific consequences", which the programmer must
+be aware of.
+
+# Detailed design
+
+Add a new function attribute to the language, `#[naked]`, indicating the
+function should have prologue/epilogue emission disabled.
+
+Because the calling convention of a naked function is not guaranteed to match
+any calling convention the compiler is compatible with, calls to naked functions
+from within Rust code are forbidden unless the function is also declared with
+a well-defined ABI.
+
+The function `call_foo` in the following code block is an error because the
+default (Rust) ABI is unspecified and as such a programmer can never write code
+in `foo` which is compatible:
```

Comment:
This will be hard to enforce given how the type system works. In particular, the type of `foo` is `extern "Rust" fn()`, and so writing `let x = foo;` would be legal, and then one could later write `x()`. So I think what would be better is to have a kind of "unknown" ABI that one can use for such cases:

```
extern "unknown" fn foo()
```

The type of `foo` is then `extern "unknown" fn()`, and we would just prohibit calling functions with that ABI.